diff -up libheif-1.20.1/tests/tai.cc.orig libheif-1.20.1/tests/tai.cc
--- libheif-1.20.1/tests/tai.cc.orig	2025-07-02 13:55:11.000000000 +0200
+++ libheif-1.20.1/tests/tai.cc	2025-07-06 16:11:44.781857515 +0200
@@ -36,8 +36,8 @@ TEST_CASE( "image-tai" )
   std::string filename = get_tests_output_file_path("tai-1.heic");
 
   heif_image* img = createImage_RGB_planar();
-  heif_encoder* enc = get_encoder_or_skip_test(heif_compression_HEVC);
   heif_context* ctx = heif_context_alloc();
+  heif_encoder* enc = get_encoder_or_skip_test(heif_compression_HEVC);
 
   heif_image_handle* handle;
   err = heif_context_encode_image(ctx, img, enc, nullptr, &handle);
